dbms/src/Interpreters/Expand.cpp

```diff
@@ -0,0 +1,240 @@
+            ///      2  2       1   +  replicate_group2        for b, it's 1, we should pick and set:
+            ///      2  2       2   +                              replicate_group_rows[1].b = null
+            ///    -----------------+
+            for (size_t i = 0; i < origin_rows; ++i)
+            {
+                // for every original one row mapped N rows, fill the corresponding group set column as null value according to the offset.
+                // only when the offset in replicate_group equals to current group_offset, set the data to null.
+                // eg: for case above, for grouping_offset of <a> = 0, we only set the every offset = 0 in each
+                // small replicate_group_x to null.
+                for (UInt64 j = 0; j < replicate_times_for_one_row; ++j)
+                {
+                    if (j == grouping_offset)
+                    {
+                        // only keep this column value for targeted replica.
+                        continue;
+                    }
+                    // set this column as null for all the other targeted replica.
+                    // todo: since nullable column always be prior to computation of null value first, should we clean the old data at the same pos in nested column
+                    auto computed_offset = i * replicate_times_for_one_row + j;
+                    cloned_one->getNullMapData().data()[computed_offset] = 1;
+                }
+            }
+            block.getByName(grouping_col).column = std::move(cloned);
+        }
+        // finish of adjustment for one grouping set columns. (by now one column for one grouping set).
+    }
+    block.insert(ColumnWithTypeAndName(std::move(added_grouping_id_column[0]), std::make_shared<DataTypeUInt64>(), std::move("groupingID")));
+    // return input from block.
+}
+
+bool Expand::isInGroupSetColumn(String name) const
+{
+    for (const auto & it1 : group_sets_names)
+    {
+        // for every grouping set.
+        for (const auto & it2 : it1)
+        {
+            // for every grouping exprs
+            for (const auto & it3 : it2)
+            {
+                if (it3 == name)
+                {
+                    return true;
+                }
+            }
+        }
+    }
+    return false;
+}
+
+const GroupingColumnNames & Expand::getGroupSetColumnNamesByOffset(size_t offset) const
+{
+    /// currently, there only can be one groupingExprs in one groupingSet before the planner supporting the grouping set merge.
+    return group_sets_names[offset][0];
+}
+
+std::set<String> Expand::getAllGroupSetColumnNames() const
+{
+    std::set<String> name_set;
+    for (const auto & it1 : group_sets_names)
+    {
+        // for every grouping set.
+        for (const auto & it2 : it1)
+        {
+            // for every grouping exprs
+            for (const auto & it3 : it2)
+            {
+                name_set.insert(it3);
+            }
+        }
+    }
+    return name_set;
```

Comment:
how about caching the name_set in `Expand` and it can used by `isInGroupSetColumn` too.

Reply:
make sense